Source/monster.h

```diff
@@ -165,6 +165,7 @@ struct CMonster {
 };
 
 struct Monster { // note: missing field _mAFNum
+	int monsterId;

```

Comment:
This makes the monster struct larger for no reason, I'd rather we don't do it like this.

Reply:
@glebm would `std::distance` be an option instead?

Reply:
I am not sure how `std::distance` works (not a C++ dev) but wouldn't it need to loop over `Monsters` until it finds the index? That would be much slower than just retrieving by index I think

Reply:
`std::distance` is constant-time on random access iterators and pointers.

Still, if you need the ID, just pass the ID IMO.

Reply:
`std::distance(first, last)` is "implementation defined" (magic :D) depending on what the types of `first` and `last` are. With `Monsters` being a simple c-style array and `this` being a pointer the implementation *should* be a subtraction of one pointer from another. If monsters ever changes to a `std::list` then we'd be forced to use bidirectional iterators and walk the chain (assuming you can even get an iterator to the current element without already walking the chain).

for example msvc uses the following definition:
```cpp
template <class _InIt>
_NODISCARD _CONSTEXPR17 _Iter_diff_t<_InIt> distance(_InIt _First, _InIt _Last) {
    if constexpr (_Is_random_iter_v<_InIt>) {
        return _Last - _First; // assume the iterator will do debug checking
    } else {
        _Adl_verify_range(_First, _Last);
        auto _UFirst             = _Get_unwrapped(_First);
        const auto _ULast        = _Get_unwrapped(_Last);
        _Iter_diff_t<_InIt> _Off = 0;
        for (; _UFirst != _ULast; ++_UFirst) {
            ++_Off;
        }

        return _Off;
    }
}
```
Pointers meet the criteria for random access iterators, so the simple case gets picked.

Reply:
That's quite cool! I am learning a lot 😄

Reply:
Let me know which solution you think it's best and I can change the MR.

Reply:
You can still have `monsterId` and not make the struct any larger. `uint16_t` is sufficient as its type, and you can place it where the compiler would add padding otherwise, e.g. before `pathcount`.

Reply:
Come to think about it, there are a few tens of monsters per level, so we are haggling over 100-200 bytes maybe. It seems a bit silly.

Reply:
>  there are a few tens of monsters per level

The global `Monsters` array is currently at 56 KiB!

I'm of the "pass the ID if you want the ID" opinion not just for the size reasons but also because it's good to explicitly communicate that the function needs the ID specifically (everything else should pass by reference but that's an ongoing parallel refactoring).

Reply:
`MAXMONSTERS` = 200. int = 4 bytes, so 800 bytes i guess.

Reply:
> Very few functions actually need the ID.

The problem is that the few functions are at the end of a call chains leading to all of them having to pass the ID down the stack.

Reply:
> The problem is that the few functions are at the end of a call chains leading to all of them having to pass the ID down the stack.

If there are a lot more functions than what's addressed in this PR then this is the implementation that I'd prefer:

```c++
size_t Monster::getId() const { return std::distance(&Monsters[0], this); }
```

Less stuff to keep track of.

Reply:
The extra memory usage was an afterthought. `struct Monster` still has padding where you can place an `uint16_t`, without enlarging it.

Reply:
I think this is actually the case for the 60 renaming monster related functions. the of which 17 appears to use the ID.

Reply:
I'd place it outside of the class, i.e..:
```C++
size_t getMonsterId(Cmonster *monster) const { return std::distance(&Monsters[0], monster); }
```

Reply:
The std::distance proposal seems a better solution than adding a new field to the struct. It’s less state to keep around and state adds complexity.

Reply:
Bump.

I updated the PR. I removed the m`monsterId` from the struct and implemented the `GetId` method.

I play-tested the PR and it seems to be working